Take full control of the actions with the most sophisticated approaches- Quota Supervisor

uluQulu opened the pull request:
If you ask why I did open this again?
Hm, 🤔

I started to work on QS again specially after @the-unknown post in #2617.
And now with a little bit more python skills (_than the start of the year when I introduced it first_) I did rewrote the core of it- the **realtimeness** of it using normal methods 👈🏼 😂.
- Well why normal methods?
Cos earlier it was based on hardcore `SQL` queries, yes every bit of data was taken from the **DB**: _settings_ & _records_ **both**.

Now, it uses the very helpful `Settings` class (_thanks to @lucanello_). How?
- It loads the records from the DB **only** at the start of the program to that global class. That's it!

As a result using global classes helped to store global data and be used throughout the source code and YET preserve the realtime service with no leaks.
>Earlier I couldn't realize it **COS** I HAVE not read about classes 🤓 from my programming book and hope will read soon to write much efficient code **AND** using `Settings` class inside InstaPy I learnt it is possible..

Well, else than that core part of the mechanism- _switch from full time `SQL` querying to using simple `class`_, yes, else than that, there is no any difference in the QS.
Just, by my upgraded python skills I have refactored lots of lines, simplified some expressions, catched exception much better 😋, enhanced visually etc., 

And in the user side there is almost no any noticeable change vs the initial release.


I think I have covered every detail for the user in the [README](https://github.com/uluQulu/InstaPy/blob/master/README.md#quota-supervisor) but if you have a question, feel free to mention.


Well. Why this PR is huge?
I believe in doing one thing and doing it good. So when I was writing new tools there were blocks needed a touch I couldn't pass without enhancing it.

Here I will talk about the **BIG** changes coming with this PR:
+ **Add**: Quota Supervisor
+ **Refactor**: now there is only one unfollow engine- `unfollow_user` - tracks to unfollow ["profile", "post", "dialog"]
+ + Fixed big bug: when encounters an unavailable user broke the loop - see the usage of `emergency_exit`
+ + Enhanced: unfollowing procedure by ID
+ **Refactor**: now there is only one unfollow engine- `follow_user` - tracks to unfollow ["profile", "post", "dialog"]
+ **Visual**: changes off PyCharm inspections, PEP-8 style violations and my considerations for InstaPy style

There are lots of minor fixes I can't note all and hope you will manage to understand them 😂

What users will notice?
- **recordActivity.json** useful file will now store records in hourly basis.

I think this will be my last work shared cos soon I will be concentrated on extreme studies and then who knows 😉

-------

Cheers 😁
-------

sionking commented:
Hi it look great and clean, just make it optional please :) I will test it.

the-unknown commented:
you are amazing! testing it right away!!!

uluQulu commented:
@the-unknown 
Have you got the exact files as in my fork which this PR is based on?

If so, please add these debug lines into the load_records definition and let me know what it shows
```python
    if daily_data:
        ordered_data = {today: {}}
        zx = daily_data   # ADD THIS LINE
        print("\n\nfetched daily data:\t{}\n\n".format(zx))   # ADD THIS LINE
```

----------

Cheers 😁
----------

the-unknown commented:
@uluQulu 
I have made a full and clean clone of your InstaPy fork, found on your github, which was updated 10h ago. 

This is the debug output:

`fetched daily data:     [<sqlite3.Row object at 0x7f2812735730>, <sqlite3.Row object at 0x7f2812735af0>, <sqlite3.Row object at 0x7f2812735710>] `

uluQulu commented:
@the-unknown 
_`fetchall()` return type seems to be inconsistent_ :S

Can you replace previous debug line with
```python
        print("fetched daily data: ", daily_data)   # ADD THIS LINE
```
so that we can see the tuples inside that list rather than row objects

--------
Cheers 😁
-------

the-unknown commented:
@uluQulu 

this is the result:
`('fetched daily data: ', [(1, 0, 0, 0, 0, 4, u'2018-08-26 09:14:59'), (1, 0, 0, 0, 0, 2, u'2018-08-26 10:59:57'), (1, 0, 0, 0, 0, 1, u'2018-08-26 11:00:04'), (1, 0, 0, 0, 0, 3, u'2018-08-26 12:34:04')])`

uluQulu commented:
It is very weird, @the-unknown.
Do you still get that error?

the-unknown commented:
@uluQulu
Yes I do.

timgrossmann commented:
@uluQulu Well, this is huge! You're truly incredible.
This will make a ton of InstaPy users very happy 😮 🙌 

Love it, thank you very much @uluQulu

timgrossmann commented:
@sionking @uluQulu What do you guys think is the best way to merge the recent PR's? 
We most likely will get quite some conflicts on the PR's, so should we merge this one first since it's the biggest one?

sionking commented:
@timgrossmann absolutely not, this big PR need to be tested, I think merging the smaller one including core fixes and improv are more urgent.

uluQulu commented:
Thanks a bunch, @timgrossmann! I really enjoy writing for **InstaPy** 🙂

Cos of it changed refactors `follow` and `unfollow` engine last days were spent to test it in all of the features. Testing QS took not much time.
I mean I have tested it before put on internet, tested it in python `2.7.15` and `3.6.3` locally.

Cos of this is my PR it is up to you to merge it. But I will tell a few bits.

+ @sionking your live PR is gonna raise lots of conflicts with this, also I saw duplicate function (_of `confirm_unfollow` in here_) in there which is gonna be removed by will be removed with PR.

+ And in case of there will be other PRs changing the blocks revised in this PR, it will be look like a war in the code rather than a conflict 🔨

But I also don't want it be merged with an issue, and right now, @the-unknown has reported a very weird bug I have not solved yet. Let me solve it then decision is yours.

---------

Cheers 😁
----------

uluQulu commented:
@the-unknown 
Can you do this change to find the issue?

**open** up `quota_supervisor` file and **add** this `try`/`except` block lines before `hour = hourly_data[-1][-8:-6]` line
```python
            try:   # ADD THIS LINE
                print("let's go!")   # ADD THIS LINE
                print("daily data: ", daily_data)   # ADD THIS LINE
                print("hourly_data: ", hourly_data)   # ADD THIS LINE
                print("record time: ", hourly_data[-1])   # ADD THIS LINE
                print("hour: ", hourly_data[-1][-8:-6])   # ADD THIS LINE
                print("success!")   # ADD THIS LINE
            except:   # ADD THIS LINE
                raise   # ADD THIS LINE
            hour = hourly_data[-1][-8:-6]
```

------

Cheers 😁
------

the-unknown commented:
@uluQulu 
This is the complete result:

OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
INFO [2018-08-26 22:56:06] [username]  Session started!
ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo


..................................................................
INFO [2018-08-26 22:56:43] [username]  Logged in successfully!
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
INFO [2018-08-26 22:56:46] [username]  Tag [1/10]
INFO [2018-08-26 22:56:46] [username]  --> ilovetravel
('fetched daily data: ', [(1, 0, 0, 0, 0, 4, u'2018-08-26 09:14:59'), (1, 0, 0, 0, 0, 2, u'2018-08-26 10:59:57'), (1, 0, 0, 0, 0, 1, u'2018-08-26 11:00:04'), (1, 0, 0, 0, 0, 3, u'2018-08-26 12:34:04'), (1, 0, 0, 0, 0, 3, u'2018-08-26 16:23:32'), (1, 0, 0, 0, 0, 3, u'2018-08-26 22:56:16')])
let's go!
('daily data: ', [(1, 0, 0, 0, 0, 4, u'2018-08-26 09:14:59'), (1, 0, 0, 0, 0, 2, u'2018-08-26 10:59:57'), (1, 0, 0, 0, 0, 1, u'2018-08-26 11:00:04'), (1, 0, 0, 0, 0, 3, u'2018-08-26 12:34:04'), (1, 0, 0, 0, 0, 3, u'2018-08-26 16:23:32'), (1, 0, 0, 0, 0, 3, u'2018-08-26 22:56:16')])
('hourly data: ', (1, 0, 0, 0, 0, 4, u'2018-08-26 09:14:59'))
Traceback (most recent call last):
  File "./bot.py", line 42, in <module>
    session.like_by_tags(my_hashtags, amount=60, media=None)
  File "/home/username/InstaPyQS/instapy/instapy.py", line 1436, in like_by_tags
    self.logger)
  File "/home/username/InstaPyQS/instapy/like_util.py", line 200, in get_links_for_tag
    update_activity()
  File "/home/username/InstaPyQS/instapy/util.py", line 160, in update_activity
    quota_supervisor("server_calls")
  File "/home/username/InstaPyQS/instapy/quota_supervisor.py", line 44, in quota_supervisor
    quota_state = controller(job)
  File "/home/username/InstaPyQS/instapy/quota_supervisor.py", line 52, in controller
    load_records()
  File "/home/username/InstaPyQS/instapy/quota_supervisor.py", line 339, in load_records
    print("record time: ", hourly_data[-1])   # ADD THIS LINE
IndexError: tuple index out of range

uluQulu commented:
@the-unknown 
What is your full python version really?
`python --version` works in Windows, I got python 2.7.**15**

the-unknown commented:
@uluQulu 
I just upgraded to python 2.7.12 and it seems to work now.

uluQulu commented:
Hey @the-unknown 
Although it is great to use a newer version of python (_or any other software_), it is best to provide a platform independent solution.

It was a curiosity for me so I installed python 2.7.6 into my machine locally 😁
So the thing is,
```
for hourly_data in daily_data:
    print hourly_data   # 👉🏼 <sqlite3.Row object at 0x0000000002383650>
    hourly_data = tuple(hourly_data)   # convert (_or re-convert_) to tuple
    print hourly_data   # 👉🏼 (1, 0, 0, 0, 0, 1, u'2018-08-27 18:03:16') 
```

--------

Cheers 😁
---------

uluQulu commented:
I have investigated it further,
data returned by a `fetchall()` from selected rows is a list and it has **tuple-able** row objects inside it, okay.
The problem is with the backwards indexing:
zx = (1, 2, 3, 'a', 4343, u"asdasd", u"2018-10-09")
zx[-1] will raise that IndexError whilst zx[7] (_all other positive `index'es_) will return properly.
I think it is a sqlite3 module support rather than a python issue which did could not handle backwards indexing...

cool-acid commented:
`plyer` is missing from `setup.py`

uluQulu commented:
Hey @cool-acid 
Nice catch but if you noticed I added it to `requirements.txt` WHERE it should be.

But after your post I checked out **setup.py** and saw that there is a duplicate of **requirements.txt** (_only `GitPython` was not there which was added recently_)
The thing is, the usage of `install_requires` (_inside **setup.py**_) and **requirements.txt** (or any other requirement file, such as- **requirements_extras.py** or **requirements-development.py**).

Well that does not mean you can load **requirements.txt** into `install_requires` meta-data,
it is just.
`install_requires` should only contain the dependencies of the software bundle.

In my opinion and in **InstaPy** case, packages like,
**emoji** - will be used only user have emojis in strings
**clarifai** - will be user if user enables clarifai
**pyvirtualdisplay** - will be used for VPS session
**GitPython** - will be used only for something useful I don't know 😁
**plyer** - will be used only for toast notifications if `notify_me` param is enabled

What I mean is those kind of packages does not have to be inside `install_requires` cos they are not dependancies. But **selenium** is, so is **requests** package,..

But well, those extra packages must be located inside `extras_require` meta-data, right. The problem is InstaPy's userbase is quite beginners of python and they will end up with confusion if it was so, cos
in case of the perfect state of requirements, it should be.

**requirements.txt** | **requirements-full.txt**, **requirements-test.txt**
-------------------|----------------------------------------------
`install_requires`=["selenium", "requests", "setuptools"] | `extras_require`={"**full**":["emoji", "clarifai", "GitPython", "pyvirtualdisplay", "Pillow", "wheel", "plyer"], "**test**":["pytest", "tox"]}


And during installation user must specify the extras desired, e.g., in the project root folder
```powershell
pip install -e .
```
command will install `install_requires` (**requirements.txt**) only [_typical installation_]
AND
```powershell
pip install -e .["full"]
```
will install `install_requires` (**requirements.txt**) & `extras_require["full"]` (**requirements-full.txt**) both
but not `extras["test"]` (**requirements-test.txt**).

As you see it is a bit complicated for the user, but with a good documentation it can be solved **in the future** and for now I put every requirement as a dependency into the `install_requires` parameter to make things easier for users as of today.
Added `extras["test"]` for the future and is a must-have functionality.

Added several useful meta-data to `setup()` updated some, refreshed package versions.
So here is my final words on InstaPy through PyPI standards,
- [x] Setup is 90% ready
- [x] **requirements.txt** versions may mismatch in some old platforms [_needs tests_] but is not a problem
- [] **MANIFEST.in** must get a revision but has not so serious problem

------

Cheers 😁
-------

uluQulu commented:
In the latest commit- https://github.com/timgrossmann/InstaPy/pull/2771/commits/99774d65deaf1d44660971eb95a9c66bcdd43ef2 I have solved using **new tab** at a context manager.  
Why?
We need the user `ID` and directly navigating to webpage kills dialog references & cache.
Also using followed-clicks from dialog pops up current issue- IG has a 1 time history navigation limit.

So I used this idea:
```python
# code code
if track == "dialog":
        with new_tab(browser):
            web_address_navigator(browser, user_link)
            user_id = find_user_id(browser, username, logger)
# code code
return user_id
```
+ Scrolling ~
+ Simulation follow
+ Open new tab
+ Navigate directly to userpage (to get user ID from `ProfilePage` entry) and return user ID
+ Close new tab
+ Continue scrolling ~

It is the most optimal solution I could think of.
Also `new_tab` definition will be a useful tool for usage in the future.  
It easily opens and closes the new tabs in request without a leak.

Is it efficient?
- [x] Considering that simulations happen randomly in fewer amounts, it is an efficient solution.

Could there be better solution?
- [x] Yes. If some JS programmer could kill a few event listeners it would do fine but it is also risky.

------

Cheers 😁
-------

sionking commented:
@uluQulu you will fix small problem by adding many others, don't open new tab.
maybe saving the cache to a var is better ?

uluQulu commented:
@sionking 
Go ahead and do it, that's all I could do.
For its usage, it is pretty stable in graphical and headless modes and really don't be scared of working with tabs, tons of people use it xD

-------

Cheers 😁
--------

uluQulu commented:
In the latest commit I have added **LIVE REPORT** lite-feature:
+ Attached to the `end` method to see the sessional stats while ending the session
+ Can be added to the quickstartscripts easily by `session.live_report()` line to see the live sessional stats

I have had this function for a long time in my quickstart script as a  definition, just thought it will be userful to everybody.
```erlang
INFO [2018-08-29 10:50:15] [qulu]  Sessional Live Report:
        |> LIKED 89 images  |  ALREADY LIKED: 7
        |> COMMENTED on 20 images
        |> FOLLOWED 34 users  |  ALREADY FOLLOWED: 5
        |> UNFOLLOWED 66 users
        |> INAPPROPRIATE images: 18
        |> NOT VALID users: 10
currently FOLLOWING 1001 users & has got 1000 FOLLOWERS
```

-------

Cheers 😁
-------

HCWcoder commented:
@uluQulu
Wow that is masterpiece thanks for the share!

fgisslen commented:
Great! let's merge this PR :)

the-unknown commented:
@uluQulu 

don't know how this affects the bot, but I am getting the following error message:

INFO [2018-08-31 11:41:26] [username]  Quota Supervisor: hourly likes reached quotient!  ~going to sleep 27 minutes long
        take a coffee colombia break? :>
ERROR [2018-08-31 11:41:26] [username]  Quota Supervisor: toast notifications currently isn't supported on your platform!        ~you shouldn't use `notify_me` parameter :L
        No usable implementation found!
Traceback (most recent call last):
  File "/home/username/InstaPyQS/instapy/quota_supervisor.py", line 278, in toast_notification
    ticker="To switch supervising methods, please review quickstart script")
  File "/home/username/.local/lib/python2.7/site-packages/plyer/facades/notification.py", line 53, in notify
    app_name=app_name, timeout=timeout, ticker=ticker)
  File "/home/username/.local/lib/python2.7/site-packages/plyer/facades/notification.py", line 58, in _notify
    raise NotImplementedError("No usable implementation found!")
NotImplementedError: No usable implementation found!
INFO [2018-08-31 11:41:26] [username]       ~consider installing the latest version of 'plyer' module by running
`pip install -I https://github.com/kivy/plyer/zipball/master`

the latest plyer version is of course installed.
This is python 2.7.12

the-unknown commented:
@uluQulu yes I have installed plyer using the suggested method:
`pip install -I https://github.com/kivy/plyer/zipball/master`

I installed in any possible way: one with "sudo", one with "--user".
But the error still appears. Otherwise I would not have reported this error message. ;-)
However it seems not to have any negative effect to the bot itself.

uluQulu commented:
### OTHER ADDITIONS & CHANGES NEEDS MENTION:

Here are the key points that I have to say to be KNOWN.


#### 1-) What is **`self.quotient_breach`**?  
```python
self.quotient_breach = False
```  
This little bool is just helping us to handle loops after quotient reaches peak.  
There is some logic here I will shortly tell:  
First of all quota supervisor will not allow any action any where to be happened out of the peaks. Right.  
But there are features in the program which has loops [normally]. It will be better in an example;  
`like_by_tags(amount=200)` [LIKE DEPENDANT feature]  &  `quota_supervisor(peak_likes=92)`  
+ It likes 92 images successfully [OK]  
+ The loop will continue opening new images from the fetched 200 links in the rest 108 images and QS will not any of them to be liked. Cos the likes has reach its peak.  
 
##### Why it happens so?  
QS does not break any loop, it just JUMPS actions in case of it exceeds the peak set. It is a flexibility spirit.  
But wait who would like to continue the rest 108 images and saying "_IMAGE NOT LIKED! lIKES REACHED LIMIT!_"? - I guess noone.  
Fine, here is `quotient_breach` takes the role.  
It breaks those loops for us after e.g. 7 consequent JUMPED likes. [read **jumps** in the next part]  

There is one more specification:  
Such as, take this,  
feature: **follow_user_followers**;  possible actions: [follow, unfollow (_simulated_), like, comment];  
During the run in **follow_user_followers**, any (_or all_) of those actions can exceed the peak and be jumped consecutively.  
So comes into the mind which action(s) should stop the loop in my feature?  
I have thought of this a lot and I decided to use direct logic: the core action  
So, for **follow_user_followers**, its core action- _follows_ will break the loop after consective follow jumps.  
>Most of the features have deterministic core actions although e.g. `interact_by_users` seems a bit confusing but I decided _likes_ to be its core action :)

The ideas above are generic and applies to all of the features.  
_If you wonder the chose core action of any feature, you can find it around `self.quotient_breach = True` lines_.


#### 2-) More on **`self.jumps`**  
```python
    self.jumps = {"consequent":{"likes": 0, "comments":0, "follows":0, "unfollows":0},
                  "limit":{"likes": 7, "comments":3, "follows":5, "unfollows":4}}
```
Read in previous block, `self.jumps` stores the JUMPS data;
+ consequent jumps: if any successful action takes place, then consequent jumps label will reset OR if it will reset after breaking all of the [nested or not] loop(s).
+ limit of consequent jumps [_will break loops after this limit_]: the pre-defined limits above are recommended values by me and user will not touch it

#### 3-) Why `self.internal_usage`?
```python
self.internal_usage = {"follow_by_list": {"validate":False}}
```
There are features available to use either internally [by the program itself] or by the user.
And there are lots of features that kind which are using others internally..  
Such as, **follow_commenters** feature uses **follow_by_list** feature to do _follow_ [_plus interact_] which itself also uses other feature- **interact_by_users** to _interact_,. etc.
The use case is common and large, that's where `internal_usage` variable holds the NAMES & _helpful parameters_ of the features _used internally_.   
_Read below to get better understanding_...

#### 4-) How does `feature_in_feature()` method work?
As can be seen from its name, it is smartly handing the activity flow while a feature is being used by another feature, 
```python
def interact_user_following():
    ## code code
                    with self.feature_in_feature("interact_by_users", False):
                        self.interact_by_users(person,
                                                self.user_interact_amount,
                                                 self.user_interact_random,
                                                  self.user_interact_media)
```
Above `interact_user_following` [host feature] is using `interact_by_users` [guest feature]. [Search `with self.feature_in_feature` in **InstaPy** class to see all use cases]
What are the benefits?
+ It provides quotient breach operate with no leak in efficiency;  
_if quotient breach reaches inside the guest feature and it leaves it, then it will also leave the host feature and prevent unwanted behaviours_
+ Some log messages needs to be printed only in the host mode [standalone mode rather than guest mode]
+ In certain cases users are already validated, so it will not validate them again [which costs navigating to the user's profile] - the change is efficiency
+ Use case is quite resilient

>Basically, `feature_in_feature()` is just adding the guest feature  into the `self.internal_usage` before starting guest and removes it from `self.internal_usage` after leaving the guest.  

#### 5-) `standalone` variable
Mentioned above, it tends to mean that the feature is being run in standalone mode, in other words, it is run by the user, normally.
In standalone mode, feature will print all of the messages and it will validate all of the users.
Otherwise it prints only required messages and validates users if they are not already validated.



#### 6-) Generic return type in all actions
After a big refactor on follow and unfollow engines, now, all of the actions does have a generalized form
+ likes: `like_image()` > searches for _Like_ button and tries to put like [from: post page]
+ comments: `comment_image()` > searches for _Comment_ buttons and tries to do comment [from: post page]
+ follows: `follow_user()` > searches for _Follow_ button and tries to do follow [from: dialog box; post page; profile page]
+ unfollows: `unfollow_user()` > searches for _Unfollow_ button and tries to do unfollow [from: dialog box; post page; profile page] 

>If you noticed, now, refactored `follow_user()` and `unfollow_user()` unfollow engines is not part of any loops, they are truly standalone engines.  

Right, after action engine standartization, I did generalized the return types:
Now all of the action engines return 2 total values: 
+ `state` [True, False]
+ `message` ["success", "commenting disabled", "already followed", "invalid element", "jumped", "etc."] 

> Earlier, some of them was returning [1, 0] and others [True, False], etc.

#### 7-) Edited `comment_by_locations()`
Now this feature has the commenting priority. It means it will only look for to follow only after putting a comment. 
>If user wants overwise, then there is `interact_by_users` feature which has _follows_ free of _comments_ [_likes_ can be disabled]

#### 8-) Edited `interact_by_users()`
Now it will try to comment only after putting like, cos otherwise is absolutely non-humanly.  
Yes. Thinking the users' comments are in positive or  neutral [non-negative] mood, you would comment saying something good and not liking is non-humanly.  
I do not think users would use negative comments.  
OR other possible case is sending promotion comments like "_Hey @dorothy! Checkout my page!_" on a post without liking, well this kind of comments should be avoided :D


#### 9-) Basic `TimeoutException` handling for `GET` navigation method in web_address_navigator()
It has been raised by @fgisslen who frequently faced timeouts and I inspected the log files and found out that almost all of those timeout exceptions occured during `browser.get()` activity.
Well, athough the scope of `TimeoutException` is HUGE, our usage of selenium is quite simpler for now and it is hitting only the GET navigation method for now.
That's why this basic solution- handling it inside `web_address_navigator()` function will be a good solution.
>But in future more generalized solutions can be offered if faces timeouts from different blocks else than `GET` method.


#### 10-) Small or big logical fixes [to existing fulfill expectations]
 

#### 11-) Add feature-wide action stats for some features & Improve some
_e.g._
```erlang
Finished following commenters!

# print results 
Followed: 100
Already Followed: 3   # NEWly added
Not valid users: 9

# print these below if interaction is enabled [otherwise they all will be NULL]
Liked: 40
Already Liked: 2
Commented: 9
Inappropriate: 5
```

After adding `self.already_followed` variable now **Already Followed** stats can also be seen
It is actualy mainly cos of having a generic return type from `follow_user()` follow engine:
e.g., in case of.
```python
follow_state, msg = follow_user()

if follow_state == True:   # followed
    # code code
elif follow_state == "already followed":
    already_followed += 1

```
This applies to all of other actions engines. Now cos of those return types, there is robust stability in valuing. 
>Earlier it was doing, such as,
```python
liked = like_image() 
if liked:
    # code code
else:
    already_liked += 1
```
_It was unreliable cos it can return from `like_image()` with other scenarios rather than "**already liked**", such as, "**invalid element**", etc_.

Read more of generic return types of action engines in **6**th step


#### 12-) Fix small but also big hideous bug
Unfollowing inappropriate users was not working out of inordinate logical statement!
 It was
 ```python
if reason == 'Inappropriate' and unfollow: 
```
Ealier it was just returning `Inappropriate` and after adding the lines of printing blacklisted word [implemented by me :/], it was somehow forgotten to fix that condition..
And that condition was never gonna be fulfilled cos NOW `reason` returns,
`Inappropriate! ~ contains 'disliked_word'`  OR `Inappropriate! ~ contains 'disliked_word in some_word'` 
Now is fixed after changed that conditional check to,
```python
if "Inappropriate" in reason and unfollow
```

#### 13-) Take whole `end()` method into safety
Whole of `end()` method is in the care of `interruption_handler()`.
>While deleting cookies and closing webdriver, it was printing "_Could not locate Chrome_" in case of EVERY `WebdriverException` WHICH truly is a misleading error log!

But both __deleting cookies__ & __closing webdriver__ are TWO different IMPORTANT & __problematic__ steps, that's why I changed a few lines 
+ EACH now is handled separately in their own `try`/`except` blocks rather than one- combined.  
Why separate blocks? - Cos once one the one above throws and exception, it didn't execute the other one, that is one of the reasons why they needed to be in separate blocks.   
+ Now Catches exceptions wisely

#### 14-) Relocate imports in neatly order [in the files I worked]
```python
# import built-in & third-party modules
import time
from selenium.webdriver.common.proxy import Proxy, ProxyType

# import InstaPy modules
from .database_engine import get_database

# import exceptions
from selenium.common.exceptions import NoSuchElementException, WebDriverException
```
This will improve source code strength and make things a bit easier for future maintainers. 


#### 15-) Simplify `Settings` CLASS
Especially the lines of newly added capability for OS specific chromedriver support has been highly simplified.  
>I really like the `Settings` CLASS stay simple, I don't know why :P

_Right, it is just simplification- I mean the mehcanism remains the same_. [Actually, I do tell when I change a bit in a mechanism, so do not worry :)]


#### 16-) Simplify & enhance `set_automated_followedpool()`
I looked into after @CharlesCCC added user `ID` to be also recorded and it was loading into the `set_automated_followedpool()` which I ended up refactoring ONCE I merged it into my PR WHERE it is part of Unfollow feature.
Else than simplification and structural clarity, NOW, there is a special point in there:
When user has no user `ID` in recs [old record or for some reasons], then returns its `ID` as `False` [rather than `None`], SO THAT, it does not search for it in `unfollow_user()` engine again.
>Generally `unfollow_user()` will look for the ID of the user [if it fails to unfollow by username] all the time by reading **dorothy_followedPool** CSV file which is a time consuming activity so that little point will dismiss duplicated checks :P 



#### 17-) The EXACT procedure in `unfollow()` feature after separating ["_go find Unfollow button any click it_" step] into **unfollow engine** (`unfollow_user()`) from the loops
+ Ease of control after the separation [_refactor_].
                    _e.g._, now has emergency conditions - ["shadow ban", "not connected", "not logged in"] etc **labels** to terminate the loop in `unfollow()` definition
+ Inside `unfollow_user()` engine, different tracks has the same general tasks after _successful_ or _unsuccessful_ **unfollow**
_e.g._ `delete_line_from_file()`, `update_activity("unfollows")`, etc. SO using the common code simplifies tons of lines 

#### 18-) Route more `GET` navigation lines to `web_address_navigator()` to INCREASE speed and SAVE server calls
Also in latest commits added higher level of stability to `web_address_navigator()` function:
+ Now it removes the last slash in the given link and current link then it compares 100% accurately
e.g. earlier if the provided link was "_https://www.hubahuba.com_" but current link was "*https://www.hubahuba.com/*", then it was navigating again.
+ after verifyhing it the provided link is different than current link, then it **adds the slash back** [_if it had a slash_] to the end of link COS slash at the end navigates pages much faster assuming it is a directory.  

#### 19-) Bunch of refactors into functions WHICH are being used several times

#### 20-) Simplify & revise some little bits in `update_activity()`
+ Now default action - `update_activity(action="server_calls")` is explicitly server calls [running `update_activity()` **tended** and now CLEARLY **tends** to mean: "_update server calls_"]
+ Simplify bunch of lines into `data[action] += 1`


#### 21-) Fully featured `remove_duplicates` for use in future [STILL IS NOT CURRENTLY IN USE]
`remove_duplicates(containter, keep_order, logger)`
**container**: _can be list, tuple, dict or anything iterable_; **keep_order**: _removing dups without keeping order is super fast_  
+ Currently supports lists only
+ Uses `sorted` algorithm



#### 22-) Fix program stopping bug in `delete_line_from_file()`
```python
if not os.path.isfile(filepath): 
    #in case of there is no any followed pool file yet 
    return 0
```
It happens in case of fresh installs or any other times **followedPool** CSV file does not exist


#### BONUS
##### How much time Quota Supervisor inspection take?  
Well, during that quick test I had a middle sized data in `recordActivity` table of the DB.  
And tested on a machine with `1.41` Ghz **CPU** & `6144` MB **RAM**.  
_test style_:
```python
zx = time.time()

if quota_supervisor("likes") == "jump":
    return False, "jumped"

fx = time.time()
print(fx-zx)
```
+ **server calls**:
`0.0010097026824951172`
`0.004998207092285156`
`0.0009984970092773438` 
`0.0`  # most of time

+ **likes**:
`0.000997304916381836`
`0.0010013580322265625`
`0.0010018348693847656`
`0.0`   # most of time



If you want to learn more about **QUOTIENT BREACH** design decision, check out the my [personal notes]() on that matter. 

_There can be some typing mistake or anything of that kind in this comment, I hope you get it right_. 

-------------------------

Cheers 😁
--------------------

uluQulu commented:
Thanks @timgrossmann

No problems, I decided to turn off logger messages for toast notifications [since it is not a big block to be missed..]:
+ Use it if it is available
+ Disable it [for the entire session] if it is not

Exactly, this piece changes nothing in the outer scope. Only users will have to install plyer from pip.
QS is available at request and will be used if enabled.
>If you look at those,
`if self.jumps["consequent"]["follows"] >= self.jumps["limit"]["follows"]:`
loop [nested] breakers, well since QS is not in use [not enabled] the consequent action values will remain null and will not trigger those conditional statements.

The same for `self.quotient_breach`, it is only be activated after consequent jumps exceed jump limits.

QS is really optional, believe me :D


I will shortly summarize it,
ACTION -> ACTION_ENGINE --> QS [enabled] ? [INSPECT] ---> [ACTION AVAILABLE] ? [TRY ACTION]
                       |                             ---> [ACTION NOT AVAILABLE] ? [return "jumped"]
                       |
                        -=> QS [not enabled] ? [PASS INSPECTION] ---> [TRY ACTION]
ACTIONs = ["likes", "comments", "follows", "unfollows", "server_calls"]


That's it.

It has a full controll over the action engines.
The rest of code for QS is breaking the loops [nested] after consequent jumps off the predefined limits.
*Any successful action resets consequent jumps counter
LIMITS for consequent jumps TO break the LOOPs in features: {"likes": 7, "comments":3, "follows":5, "unfollows":4}

You can change it in future, it is in the InstaPy class `__init__()`, I think those values are good for now.

TIP: Breaking nested loops is difficult rather than simple one loop. There are some method to break those nested loops and the one I used is the simplest and most clearer and quite an efficient one.

But there is more elegant solution I know but I did not inclue it cos it will indent everything below, and GIT diff will show all code as changed :D

e.g. instead of 
```
for location in locations:
    if stop_parent_loops==True:
        break

    for country in location:
            if stop_parent_loops==True:
                break

            for city in country:
                if stop_parent_loops==True:
                    break
                if stop_parent_loops==True:
                    break            

                for place in city:
                    if stop_parent_loops==True:
                        break

                    if place == "Home town":
                        print ("Hey reached home!")
                        stop_parent_loops=True
                        break
```
Although in InstaPy there are only max 3 nested loop [I remember so] I know more efficient way like this,
```python
with smart_looping() as stop_all_loops:
    for location in locations:
        for country in location:
                for city in country:
                    for place in city:
                        if place == "Home town":
                            print ("Hey reached home!")
                            raise stop_all_loops
```

so what is `smart_looping()`


```python
@contextmanager
def loop_breaker():
    class NestedBreakException(Exception):
        pass

    try:
        yield NestedBreakException

    except NestedBreakException:
        pass
```

Although python master will not like using Exceptions like this :{ I still liked that elegant style :P






Olright, I will write some super easy examples.
I merge conflicts tomorrow, let's see.

This PR was ready when I pushed.



All the commits I done are additional changes unrelated neither to QS nor this PR. Those are just some upgrades in other places or some additions in other places I did last days.
The only problem:
+ Crazy IndexError in python 2.7.6 [FIXED]
Other bug:
+ in get_following_status, changed `return "Following"` to `return True`


I will add one more commit tomorrow, fixing 2 problems with logging in - inconsistent sleep delay [users without cookies] and handle the 'Get App' offer.
There are other changes, too.

---------

Cheers 😁
-----------

timgrossmann commented:
@uluQulu Thank you for the explanation. Good to know that it's completely optional and that all the changes you did after creating the PR are just for additional fixes.

I'll take some time and look through anything once you're done with the travis-ci failure and give me the go ☺️

uluQulu commented:
Hey @timgrossmann 
Just have completed all I could do.
Noticable mentions are:
#### NEW TAB usage [reverted]
 There were a troublesome "_get user ID from dialog box_" WHERE I offered a new tab approach which worked flawlessly on mine and some other WHILST ended up with UNCLEAR bugs in other and due to having no any accurate FEEDBACK from the affected users I could not solve it THAT'S WHY I removed the phase of "_getting user ID from dialog box_" [_only during simulated follows_] and now those users' ID will not be recorded as before 🙂 and I think people will solve it in future 🙋🏼‍♂️ (_`new_tab()` is still there_)

#### expansion of `explicit_wait()`: https://seleniumhq.github.io/selenium/docs/api/py/webdriver_support/selenium.webdriver.support.expected_conditions.html
There are all of the available tracks [_expected conditions_] that can be used with the `explicit_wait()` definition.
Fine, while testing other _newer_ account which did not have a cookie loaded, it crashed at the same spot, `2` seconds of sleep was not enough to have the `username` box located in the page at `.com/login` address.
Well, it was `1` seconds not working before and now `2` seconds. I thought instead of regularly updating that second game, it is best to leave it to the `explicit_wait()` function.
Where to use `explicit_wait()` - many places.
E.g. for that login page, I did 
**1**-)
```python
    # wait until it navigates to the login page
    login_page_title = "Login"
    explicit_wait(browser, "TC", login_page_title, logger)
```
_"TC" means `title_contains` expected condition_;
Why needs to wait?
- Cos there is `username` box in the `.com/` address too [for registration] BUT we need that box in `.com/login` address to log in
**2**-)
```python
    # wait until the 'username' input element is located and visible
    input_username_XP = "//input[@name='username']"
    explicit_wait(browser, "VOEL", [input_username_XP, "XPath"], logger)
```
Well, instead of previously static 2 seconds of sleep (_which was inconsistent and should be increased to `3` or `4`_) NOW it does `explicit_wait()` with its default timeout of `66` seconds.  
How it works?
- It waits until the element is located and is visible - "VOEL" track (_`visibility_of_element_located` expected condition_) and then returns immediately.
And in case of element is not found, it will timeout after `66` seconds.
>Well in those scenarios default timeout is fine cos those actions both are **crucial**.

**3**-)
```python
def dismiss_get_app_offer(browser, logger):
    """ Dismiss 'Get the Instagram App' page after a fresh login """
    offer_elem = "//*[contains(text(), 'Get App')]"
    dismiss_elem = "//*[contains(text(), 'Not Now')]"

    # wait a bit and see if the 'Get App' offer rises up
    offer_loaded = explicit_wait(browser, "VOEL", [offer_elem, "XPath"], logger, 5)

    if offer_loaded:
        dismiss_elem = browser.find_element_by_xpath(dismiss_elem)
        click_element(browser, dismiss_elem)
```
This is a bit different and here I used a _pre-defined_ timeout of `5` seconds. Why?
- `Get App` offer does appear only on some accounts (_possibly new_) and it pops up immediately in less than 3 seconds.

I really like the usage of **waiting until expected conditions** and I think `explicit_wait()` function can be applied to many blocks to increase efficiency plus responsiveness and decrease waiting time.

#### @sionking I removed `random.randint(1, 5) == 1`
Well, since the follow button is `"Following"` [after clicking to follow] it doesn't need to re-load page and check again COS I don't think the button will revert back to the non following state after a re-load.
_Now it re-checks if the button is still "Follow" or "Follow Back" OR if encounters a "Sorry..!" page_.

I think it is done, @timgrossmann and I want to thank you for the expression of faith in my
abilities 🙋🏼‍♂️ 
Thanks to all of testers and feedback providers 👏🏼
I hope it will be a worthy contribution to **InstaPy**.

----

Cheers 😁😁
-----

sionking commented:
@uluQulu never Liked blocked  ?
It will mark the heart in red but after reload it is empty again,
I never experienced it with follow but I guess it can be the same.

sionking commented:
@uluQulu thank you for everything.
But I think you make this program a little too complicated.
Once you will no longer be here, who will understand what you did to solve issues?
This is a place to learn python but I think using most of it powerful features is a bad idea for a community  application. most of the programmers here are inexperienced.
I hope you will stay with us for as long as those modules are tested and working.

Thanks again.

sionking commented:
Regarding login, I found it the simplest to just wait the user input when we are on login page.
It done once, lets do it manually (and also not saving the password in plain text for security reasons)

fgisslen commented:
@sionking it's not a good idea to wait for user input for login, just sayin.

sionking commented:
@fgisslen after cookies are created you will not wait again. done once.
So it easy to handle unexpected results like suspicious login attempt ...

fgisslen commented:
@sionking yes, for suspicious login attempt it's great. But if you mean for password it's not.

uluQulu commented:
@sionking, so the PR is good for you as is? @timgrossmann, review? (after semester, of course :])

uluQulu commented:
@sionking 
>I never experienced it with follow but I guess it can be the same.

I think if you think if it really is as you GUESS, then it will be so easy to add that support:
```python
if not following:
```
line will be
```python
if not following or random.randint(1,5)==1:
```
That's it.
But it must be verified because having a reload on ~20% of follows will cost lots of requests to the server and these days requests are so expensive 😛

>This is a place to learn python 

Heh, I hope it is.

>But I think you make this program a little too complicated.

Right, Photoshop is also complicated but people prefer it over Gimp 😁

>Once you will no longer be here, who will understand what you did to solve issues?

I have always been available to solve the problems out of the lines I touched.  
And in any case I wrote those long explanatory comments for quick reference in design decisions 🙂

>but I think using most of it powerful features is a bad idea for a community application. most of the programmers here are inexperienced.

Well since I do not yet know about powerful features how can I write a one? 😂
Although, having powerful features is good and it will force people to learn them, in case of they are interested in this program.

>I hope you will stay with us for as long as those modules are tested and working.

Sure, all of my written lines have a full guarantee from my side [as long as I do have an internet access 💚].  
Do you really think some module from this PR is not working? I really wonder which one it is.

For your ideas on login,
I think you did not read my post on `explicit_wait()`. I said there are two username boxes and one in `.com/` [for registration] the other in `.com/login` for logging in and there needs to be some sleep time to wait page correctly load to `Login` page and then look for the username box AND `explicit_wait()` does it and other similar stuff SMARTLY and EFFICIENTLY.

Here there is a word "Automation" in the description of **InstaPy** and you offer to add an `input()` line to kill that automation 😳
I really agree on password and username being somehow unsafe and I saw some people post personal information inside quickstart script when requested.  
That's why that confidentiality must be handled properly, e.g., having a separate file to store sensitive information can be added in future and other files need it just loads it using `open()`.

If bypassing suspicious attempt is not working properly, it can be fixed, it is just a matter of talented hands 👨🏼‍💻👩🏼‍💻 👈🏼 😉

-------

Cheers 😁
------

fgisslen commented:
I have tested and it works fine for me.

fgisslen commented:
I think I found an error in QS:

INFO [2018-09-04 13:57:27] [xxx]  Quota Supervisor: hourly likes reached quotient!        ~going to sleep 5 minutes long
        take a black tea break? :>
INFO [2018-09-04 13:57:29] [xxx]  Sessional Live Report:
        |> LIKED 0 images  |  ALREADY LIKED: 0
        |> COMMENTED on 0 images
        |> FOLLOWED 0 users  |  ALREADY FOLLOWED: 0
        |> UNFOLLOWED 0 users
        |> INAPPROPRIATE images: 0
        |> NOT VALID users: 0
currently FOLLOWING 683 users & has got 1706 FOLLOWERS


OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
INFO [2018-09-04 13:57:29] [xxx]  Session ended!
oooooooooooooooooooooooooooooooooooooooooooooooooooooo



Traceback (most recent call last):
  File "5b8825d93e1f010014e23a2d.py", line 25, in <module>
    session.like_by_tags(['xxx'], amount=100)
  File "/home/fredrik/InstaPy/instapy/instapy.py", line 1492, in like_by_tags
    self.logfolder)
  File "/home/fredrik/InstaPy/instapy/like_util.py", line 519, in like_image
    if quota_supervisor("likes") == "jump":
  File "/home/fredrik/InstaPy/instapy/quota_supervisor.py", line 44, in quota_supervisor
    quota_state = controller(job)
  File "/home/fredrik/InstaPy/instapy/quota_supervisor.py", line 73, in controller
    toast_notification(notify, "sleep", job, interval)
  File "/home/fredrik/InstaPy/instapy/quota_supervisor.py", line 278, in toast_notification
    ticker="To switch supervising methods, please review quickstart script")
  File "/usr/local/lib/python3.6/dist-packages/plyer/facades/notification.py", line 47, in notify
    app_name=app_name, timeout=timeout, ticker=ticker)
  File "/usr/local/lib/python3.6/dist-packages/plyer/platforms/linux/notification.py", line 34, in _notify
    session_bus = dbus.SessionBus()
  File "/usr/lib/python3/dist-packages/dbus/_dbus.py", line 211, in __new__
    mainloop=mainloop)
  File "/usr/lib/python3/dist-packages/dbus/_dbus.py", line 100, in __new__
    bus = BusConnection.__new__(subclass, bus_type, mainloop=mainloop)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 122, in __new__
    bus = cls._new_for_bus(address_or_type, mainloop=mainloop)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.Spawn.ExecFailed: /usr/bin/dbus-launch terminated abnormally without any error message

uluQulu commented:
@fgisslen 
Does it always happen or sometimes (_how frequently?_)?

fgisslen commented:
@uluQulu will try again and see

fgisslen commented:
@uluQulu it seems like it happens all the time.

uluQulu commented:
@fgisslen 
Now after https://github.com/timgrossmann/InstaPy/pull/2771/commits/1813963a830c1b8bb2a474381bc395cc39e00a30 it should solve your problem 🤝🏼

fgisslen commented:
on line 299 in instapy.py , could you just cast port with str(port) , I don't want to do PR about that little thing.

uluQulu commented:
@fgisslen 
>on line 299 in instapy.py , could you just cast port with str(port) , I don't want to do PR about that little thing.
I think you really should PR it.

uluQulu commented:
@fgisslen 
It is a general bug and is not related with the changes in this PR.

But remove that redundant `body_elem.send_keys(Keys.HOME)` line from **like_util.py** file

-------

Cheers 😁
--------

timgrossmann commented:
@uluQulu Just tested it today and it seems to be working very well!
Both (still) without using it and when it is activated.

Also looked through the code, good that you wrote extensive documentation here in the PR, that helped a lot 😉 

Thank you for this.
@sionking I will merge this now, should there be any big errors, I'll revert the merge and we'll just PR this again.

Thanks guys.

sionking commented:
@timgrossmann @uluQulu 
For those ppl that relay on Instapy, those **big** PR are disaster.
Trying to merge it to my code will take a whole day !
I think as a community application we need to keep on small PR, otherwise it will waste a lot of ppl time trying to merge it.
Also, I can't run my users on new application, without massively testing it, for small changes I can do that, but here it is just to much.

timgrossmann commented:
@sionking I can absolutely understand this and I'm really sorry to hear that.
It's absolutely true that PR's like this one should be avoided. I also do understand that "changes" like the addition of QS needs changes as big as this one to work correctly.

Hopefully in the future we can stick to small, in the best case, "tiny" PRs that only need very little testing.
Hopefully you'll be able to still operate InstaPy for your clients.
I hope you can understand that InstaPy, mostly, is thought to be used by people for their own private accounts which means that they don't have any high responsibility for others, I'm sorry to hear that for you (and probably some others) this is the case.

Maybe it, at last, it's time to reintroduce a second branch which covers all the bigger updates to InstaPy while keeping the master branch simply on a working version with only fixes if something breaks rather than additions for now.

sionking commented:
@timgrossmann 
Even big changes can be done modulated.
This change just added a bunch of stuff to the instapy.py file and turning up side down the logic we used to have. big changes ? **modulated**.

uluQulu commented:
@sionking 

>For those ppl that relay on Instapy, those big PR are disaster.

I have always advice for people of this repo to not modify **staged** files cos it makes difficulties while merging but I see that you have too many customizations which should be avoided. Try to keep modifications in new **unstaged** files.

>Trying to merge it to my code will take a whole day !

It could. Really sorry for that 😟

>I think as a community application we need to keep on small PR, otherwise it will waste a lot of ppl time trying to merge it.

Small PR is always good, no one argues that.
But it will definitely not **waste** people's time trying to merge it cos it brings a big set of features and good revisions.  
Consider that you have a pickle-based customizations to control activity flow but ppl does not have it AND they deserve a good supervisor.

>Also, I can't run my users on new application, without massively testing it, for small changes I can do that, but here it is just to much.

Right, I agree as @timgrossmann said. I still also did not hear of any BUG or FEEDBACK out of your tests 🤓 WHO promised to test for the first time.
But good that there were other people who did try it out. 🙌

>Even big changes can be done modulated.

I agree that some of the UNRELATED changes could be shared in separate PRs, like `explicit_wait()` BUT I am a fast coder and I can't work with branches that good so included all changes in this PR.
Also note that those UNRELATED changes are just so small and I am not gonna discuss on all of the bound between the RELATED changes cos I wrote in above comments of those bounds.
_If you would like to hear, let me know to write it again_.

>This change just added a bunch of stuff to the instapy.py file and turning up side down the logic we used to have. big changes ? modulated.

Which logic has been turned upside down?
But well, if there is a wrong or uncomplete logic it really has to be corrected.

@timgrossmann, adding a test branch sounds amazing and it also sound scary in case of management which is done by you most of time 😄
+ In case of management difficulties of branches it will slow down the merges and badly affect the productivity from contributors...
+ You should consider that here are no any paid testers, just people who are interested in the PR tests it for personal usage and collaborators/contributors who has free time to try it.
+ Etc.

-------

Cheers 😁
-----
